### PR TITLE
Create Structure action

### PR DESCRIPTION
**Pull request by tokejepsen**
This is a front end action for creating directories and copying files.

By publishing an event and getting the directories and files to copy, this action stays independent of the file structure code.

### CODE REVIEW

**Review by iwootten — DISMISSED** [review]
Could you elaborate on the description on how this is used and the motivation? A screenshot of it in use might be good.

It seems a very good example of something that requires thorough unit tests. I'd break the core logic out here and test each of the assumptions being made. This is the approach I've been working using for new behaviour.

**Comment by tokejepsen** [comment]
Sure :) You can read about the setup required in the README for the action.

The motivation was to have a common action across studios that people can use for creating directories and copying files, to make up a project structure.
The reason for delegating what those directories and files look like to a second event handler, is because each studio can have their own implementation for handling file structures.

jul 27 2017 9-09 am

**Comment by tokejepsen** [comment]
@iwootten what do you think of this action?

**Comment by iwootten** [comment]
Honestly, I think it needs tests.

**Comment by tokejepsen** [comment]
Alright, lets talk about tests then.
Obviously we can't test the resulting hook in the cloud, but we could as you say test the assumptions and code leading up to the hook in the cloud.
Now as for the testing setup, I'm imagining a test file for each hook in their folders; "ftrack-hooks\create_structure\test_create_structure".
I'll be using pytest, Appveyor and Travis to test on Windows and Linux. Travis is yet to support Python natively on their builds.

What do you think about this?

**Comment by tokejepsen** [comment]
As we can't currently mock Ftrack the only code in this action is this:
```python
        for directory in event["data"]["directories"]:
            if not os.path.exists(directory):
                print 'Create directory: "{0}"'.format(directory)
                os.makedirs(directory)

        for src, dst in event["data"]["files"]:
            if not os.path.exists(dst):
                print 'Copy "{0}" to "{1}"'.format(src, dst)
                shutil.copy(src, dst)
```
Is this what you think needs testing?

**Comment by iwootten** [comment]
Yes :) - there's also the example code given in the readme. 

With regard to where the tests live - I don't see why they wouldn't live under a tests folder in the root of the project? The methods you break out could live in the existing ftrack_hooks folder.

With regard to the test setup - that sounds pretty complex for a few unit tests, how about starting by invoking them at the commandline and go from there?

**Comment by tokejepsen** [comment]
> there's also the example code given in the readme.

I don't see that code being used in the repository or the ftrack_hooks module. Do you mean using the code in the readme to test the code?

> With regard to where the tests live - I don't see why they wouldn't live under a tests folder in the root of the project? The methods you break out could live in the existing ftrack_hooks folder.

I guess if the code we break out live the ftrack_hooks module, then we just have to test the ftrack_hooks.

> With regard to the test setup - that sounds pretty complex for a few unit tests, how about starting by invoking them at the commandline and go from there?

Do you mean for testing the Ftrack part of the code?
You would definitely be able to invoke them from the commandline. Should be as simple as
```bash
cd /path/to/[ftrack-hooks repository]
pytest
```
Its more to have some auto testing on PRs which are visible for others.

**Comment by iwootten** [comment]
> I don't see that code being used in the repository or the ftrack_hooks module. Do you mean using the code in the readme to test the code?

Yeah, so you can create a few unit tests for creating and copying directories/files based around your example code.

> I guess if the code we break out live the ftrack_hooks module, then we just have to test the ftrack_hooks.

Yep exactly. Having common code live in the module and adding tests for it is the approach I'm working toward to the work I'm doing at Bait.

> Do you mean for testing the Ftrack part of the code?

> Its more to have some auto testing on PRs which are visible for others.

No, more generally - if you're willing to take the time to set up that then fair enough, but I don't think it should be part of this PR. Maybe write the tests as part of this PR and create another to add the more complex testing setup you're talking about.

**Comment by tokejepsen** [comment]
> No, more generally - if you're willing to take the time to set up that then fair enough, but I don't think it should be part of this PR. Maybe write the tests as part of this PR and create another to add the more complex testing setup you're talking about.

Sure, sounds like a good plan.
Just to be clear, you are not expecting to test the Ftrack part of the code?

**Comment by iwootten** [comment]
> Just to be clear, you are not expecting to test the Ftrack part of the code?

I'm just saying we need tests around what the use case is here, that directories and files are created as expected.

Until a ftrack mocking library is available, we can't test ftrack. I'd argue that such a library is possibly pointless from a testing point of view as we'd be mocking so much of the behaviour we wouldn't be testing anything useful.

File: create_structure/resource/hook/create_structure.py [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+
+        accepts = self.discover(
+            self._session, *args
+        )
+
+        if accepts:
+            return {
+                'items': [{
+                    'label': self.label,
+                    'varian': self.variant,
```
Comment (iwootten) [inline create_structure/resource/hook/create_structure.py]: Spelling error?

File: create_structure/resource/hook/create_structure.py [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+                'status': 'running',
+                'data': json.dumps({
+                    'description': 'Create Structure: Scanning for data.'
+                })
+            }
+        )
+        # Commit to feedback to user about running job.
+        session.commit()
+
+        try:
```
Comment (iwootten) [inline create_structure/resource/hook/create_structure.py]: Given this is the only unique 'directory creation' code for this action, wouldn't it be better to create a generic action which takes a callback/topic/description/data as arguments, so we can reuse it?

**Review by iwootten — COMMENTED** [review]
Another comment - this hook doesn't do anything without without user populating the event data. Surely the simplest use case would be creating a hook that reads a template configuration and creates a structure from that?

**Comment by tokejepsen** [comment]
> Surely the simplest use case would be creating a hook that reads a template configuration and creates a structure from that?

Originally I had a hook that did just that; https://github.com/tokejepsen/ftrack-hooks/blob/master/pipeline_plugins/resource/hook/create_structure.py, but its very studio specific.
This hook is intended as a front end for studio to feed their studio specific directories and files into.

File: create_structure/resource/hook/create_structure.py (hunk 1 above)
Comment (tokejepsen) [reply create_structure/resource/hook/create_structure.py]: Indeed, good spot :)

File: create_structure/resource/hook/create_structure.py (hunk 2 above)
Comment (tokejepsen) [reply create_structure/resource/hook/create_structure.py]: Could you elaborate a bit more on this?

Currently the intend is for the user to recognize the action "Create Structure" as something for generating directories and files setup.

**Comment by iwootten** [comment]
I guess on further evaluation, I'm not really seeing the benefit of this being a ftrack hook? If I'm having to create studio specific code to populate a data structure, there's little more work to do to create and populate those paths, as evident from what is testable. The bulk of the work here is creating an action, which I'm not sure it should be.

Another concern is whether an ftrack user should be able to do this, it seems much like more like a intial deployment task run by an admin rather than any ftrack user?

**Comment by tokejepsen** [comment]
> The bulk of the work here is creating an action, which I'm not sure it should be.

That is correct. What did you think this was originally?

> Another concern is whether an ftrack user should be able to do this, it seems much like more like a intial deployment task run by an admin rather than any ftrack user?

Good point. We could restrict who are able to see this action?

**Comment by iwootten** [comment]
> That is correct. What did you think this was originally?

I didn't think it was anything different originally - but if it's not driven by ftrack and needs to be populated, I can't see the benefit in it being a hook? I'm confused as to why we'd add 300+ lines of code over and above this just being a script for the 8 lines you want to add that an admin would run?

**Comment by tokejepsen** [comment]
> I can't see the benefit in it being a hook?

The idea was to have a cross studio hook, instead of having to implement the same action at different studios.
As you say this could be overkill compared to that you have to implement the directory and file templating logic your self.
I'm currently using it for creating a structure from lucidity templates, which looks like this: https://github.com/tokejepsen/tgbvfx-environment/blob/master/environment/FTRACK_CONNECT_PLUGIN_PATH/resources/hook/create_structure_launch.py
Maybe I should make this action more specific towards that?

**Comment by iwootten** [comment]
> The idea was to have a cross studio hook, instead of having to implement the same action at different studios.

This assumes that the structure creation needs to be an action. 

As I mentioned before, if the specific logic here to inform your action of a structure is as much work as creating the structure itself - I don't see why anyone would use it? If the structure was coming from ftrack, then fair enough - but it needs to be added to a dict, which means reading up on the ftrack action and hook api as well as this specific hook.

Personally I wouldn't want anyone creating the structure unless they were an admin and if that's the case, then it's surely part of a initialisation task performed once when first setting up a project?

I'm going to dismiss my review, so you can go ahead and proceed as you see fit, but I don't really think this belongs as a shared hook.

**Comment by tokejepsen** [comment]
> If the structure was coming from ftrack

That is the point. You create a structure from entities in Ftrack. These entities would inform how your structure would look like.

> but it needs to be added to a dict, which means reading up on the ftrack action and hook api as well as this specific hook.

Don't quite understand what you mean here?

> Personally I wouldn't want anyone creating the structure unless they were an admin and if that's the case, then it's surely part of a initialisation task performed once when first setting up a project?

I agree that it should probably be an admin that has priviliges to run this action.
Though the structure can be added to as the project goes along. For example if you add another shot to a project, which need some studio specific folders created.

**Pull request merged**